Shock/Classes/MockServer.swift

```diff
@@ -15,27 +15,54 @@ import Swifter
 
 public class MockServer {
     
-    private let port: Int
+    /// The range in which to find a free port on which to launch the server
+    private let portRange: ClosedRange<Int>
     
     private let server = HttpServer()
     
+//    private let portProvider = PortProvider()
+    
     private let responseFactory: MockHTTPResponseFactory
     
     public var onRequestReceived: ((MockHTTPRoute, CacheableRequest) -> Void)?
     
-    public init(port: Int = 9000, bundle: Bundle = Bundle.main) {
-        self.port = port
+    public var selectedPort: UInt16 = 0
+    
+    public var loggingClosure: ((String?) -> Void)?
+    
+    public convenience init(port: Int = 9000, bundle: Bundle = Bundle.main) {
+        self.init(portRange: port...port, bundle: bundle)
+    }
+    
+    public init(portRange: ClosedRange<Int>, bundle: Bundle = Bundle.main) {
+        self.portRange = portRange
         self.responseFactory = MockHTTPResponseFactory(bundle: bundle)
     }
     
     // MARK: Server managements
     
     public func start(priority: DispatchQoS.QoSClass = .default) {
-        try! server.start(UInt16(port), forceIPv4: true, priority: priority)
+        for i in portRange {
+            let proposedPort = UInt16(i)
+            do {
+                try server.start(proposedPort, forceIPv4: true, priority: priority)
+                selectedPort = proposedPort
+                loggingClosure?("SUCCESS: Opened server on port: \(i)")
+                return
+            } catch _ {
+                loggingClosure?("NOTE: Failed to open server on port: \(i), \(portRange.upperBound - i) remaining")
+                continue
+            }
+        }
+        loggingClosure?("""
```

Comment:
This should probably throw or something... but that's a big change. Thoughts?

Reply:
I guess we could add a `throw`ing version of `start`? Otherwise all the clients are going to hate us...

XCTest does that now (for `setup` and `teardown`, so there's a precedent.

Reply:
I think we should go one way or the other or it may be confusing for new or unfamiliar users (plus it means having to name it something like `startAndThrowOnError` to avoid function signature collisions).

Reply:
I think I'll add a task to revisit error handling and leave it as is for now - most of the changes we are making initially are non-breaking, so lets save the breaking ones until last 😄